pkg/agent/openflow/externalnode_connectivity.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package openflow
+
+import (
+	"net"
+
+	"antrea.io/antrea/pkg/agent/openflow/cookie"
+	binding "antrea.io/antrea/pkg/ovs/openflow"
+)
+
+const (
+	reservedFlowKey = "reservedConnections"
+)
+
+type featureExternalNodeConnectivity struct {
+	cookieAllocator cookie.Allocator
+	ipProtocols     []binding.Protocol
+	ctZones         map[binding.Protocol]int
+	category        cookie.Category
+
+	uplinkFlowCache *flowCategoryCache
+}
+
+func (f *featureExternalNodeConnectivity) getFeatureName() string {
+	return "ExternalNodeConnectivity"
+}
+
+func newFeatureExternalNodeConnectivity(
+	cookieAllocator cookie.Allocator,
+	ipProtocols []binding.Protocol) *featureExternalNodeConnectivity {
+	ctZones := make(map[binding.Protocol]int)
+	for _, ipProtocol := range ipProtocols {
+		if ipProtocol == binding.ProtocolIP {
+			ctZones[ipProtocol] = CtZone
+		} else if ipProtocol == binding.ProtocolIPv6 {
+			ctZones[ipProtocol] = CtZoneV6
+		}
+	}
+
+	return &featureExternalNodeConnectivity{
+		cookieAllocator: cookieAllocator,
+		ipProtocols:     ipProtocols,
+		uplinkFlowCache: newFlowCategoryCache(),
+		ctZones:         ctZones,
+		category:        cookie.ExternalNodeConnectivity,
+	}
+}
+
+func (f *featureExternalNodeConnectivity) hostUplinkFlows(hostOFPort, uplinkOFPort uint32) []binding.Flow {
```

Comment:
Could you add comments either before the func or before each flow to explain the behaviors?

How this is different from the uplink flows in bridging mode or Windows?

Reply:
Thanks for the explanation. Maybe we should rename it vmUplinkFlows to be clear? And add some comments saying host internal port and uplink are directly paired and no layer 2/3 forwarding flows for VMs.

Reply:
updated.